go-controller/pkg/ovn/namespace.go

```diff
@@ -0,0 +1,122 @@
+package ovn
+
+import (
+	"fmt"
+	"github.com/Sirupsen/logrus"
+	kapi "k8s.io/client-go/pkg/api/v1"
+	"sync"
+	"time"
+)
+
+func (oc *Controller) waitForNamespaceEvent(namespace string) {
+	for {
+		if oc.namespacePolicies[namespace] != nil {
+			break
+		}
+		time.Sleep(100 * time.Millisecond)
+	}
+}
+
+func (oc *Controller) addPodToNamespaceAddressSet(ns, address string) {
+	if oc.namespacePolicies[ns] == nil {
+		return
+	}
+
+	oc.namespaceMutex[ns].Lock()
+	defer oc.namespaceMutex[ns].Unlock()
+
+	oc.namespaceAddressSet[ns] = append(oc.namespaceAddressSet[ns],
+		address)
+	oc.setAddressSet(ns, oc.namespaceAddressSet[ns])
+}
+
+func (oc *Controller) deletePodFromNamespaceAddressSet(ns, address string) {
+	if address == "" || oc.namespacePolicies[ns] == nil {
+		return
+	}
+
+	oc.namespaceMutex[ns].Lock()
+	defer oc.namespaceMutex[ns].Unlock()
+
+	for i, item := range oc.namespaceAddressSet[ns] {
+		if item == address {
+			oc.namespaceAddressSet[ns] = append(
+				oc.namespaceAddressSet[ns][:i],
+				oc.namespaceAddressSet[ns][i+1:]...)
+		}
+	}
+
+	oc.setAddressSet(ns, oc.namespaceAddressSet[ns])
+}
+
+func (oc *Controller) addNamespace(ns *kapi.Namespace) {
```

Comment:
@RostyslavFridman 
The current implementation needs the namespace watcher because we need to populate the address_set for all the pods in the namespace. The annotation is being used in this implementation to make to easier to implement. We can get rid of it once it matures. Without the annotation, it is hard to figure out when to apply deny rule to a pod - specially when someone changes a network policy to add new labels or removes existing labels.